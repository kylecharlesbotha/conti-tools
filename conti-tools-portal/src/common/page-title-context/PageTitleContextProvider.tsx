import { createContext, FC, useContext, useState } from 'react';

type PageTitleContextType = {
  pageTitle: string;
  setPageTitle: (newTitle: string) => void;
};

const PageTitleContext = createContext<PageTitleContextType | undefined>(
  undefined
);

export const PageTitleContextProvider: FC = ({ children }) => {
  const [pageTitle, setPageTitle] = useState('');

  return (
    <PageTitleContext.Provider value={{ pageTitle, setPageTitle }}>
      {children}
    </PageTitleContext.Provider>
  );
};

export const usePageTitle = () => {
  const pageTitleContext = useContext(PageTitleContext);

  if (!pageTitleContext) {
    throw new Error(
      'Page Title Context cannot be used outside of Page Title Context Provider'
    );
  }

  return pageTitleContext;
};
