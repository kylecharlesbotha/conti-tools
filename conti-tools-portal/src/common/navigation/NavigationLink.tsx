import { ListItemButton, ListItemIcon, ListItemText } from '@mui/material';
import React, { forwardRef, ReactElement, useMemo } from 'react';
import {
  Link as RouterLink,
  LinkProps as RouterLinkProps
} from 'react-router-dom';
import { usePageTitle } from '../page-title-context/PageTitleContextProvider';

interface NavigationLinkProps {
  icon?: ReactElement;
  to: string;
  primary: string;
}

export const NavigationLink = ({ icon, to, primary }: NavigationLinkProps) => {
  const { pageTitle } = usePageTitle();
  // This pattern is taken from the MUI docs
  // React's useMemo is necessary here to ensure the link component doesn't re-render unnecessarily
  const renderLink = useMemo(
    () =>
      forwardRef<HTMLAnchorElement, Omit<RouterLinkProps, 'to'>>(
        (itemProps, ref) => (
          <RouterLink to={to} ref={ref} {...itemProps} role={undefined} />
        )
      ),
    [to]
  );

  return (
    <ListItemButton component={renderLink} selected={primary === pageTitle}>
      {icon ? (
        <ListItemIcon sx={{ justifyContent: 'center', paddingRight: 2 }}>
          {icon}
        </ListItemIcon>
      ) : null}
      <ListItemText primary={primary} />
    </ListItemButton>
  );
};
