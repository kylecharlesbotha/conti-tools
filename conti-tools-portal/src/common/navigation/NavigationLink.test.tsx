import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { NavigationLink } from './NavigationLink';
import HomeIcon from '@mui/icons-material/Home';
import { PageTitleContextProvider } from '../page-title-context/PageTitleContextProvider';

describe('testing the navigation link', () => {
  test('should render correctly', () => {
    const { baseElement } = render(
      <PageTitleContextProvider>
        <MemoryRouter>
          <NavigationLink to="/" primary="Home" icon={<HomeIcon />} />
        </MemoryRouter>
      </PageTitleContextProvider>
    );

    expect(baseElement).toMatchSnapshot();
  });
});
