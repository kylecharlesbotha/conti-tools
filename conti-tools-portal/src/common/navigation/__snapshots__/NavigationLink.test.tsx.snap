// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the navigation link should render correctly 1`] = `
<body>
  <div>
    <a
      class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
      href="/"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root css-1ralanr-MuiListItemIcon-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="HomeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
          />
        </svg>
      </div>
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Home
        </span>
      </div>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </a>
  </div>
</body>
`;
